src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1249,39 +1249,76 @@ private void ImportThrow()
         {
         }
 
+        private LLVMValueRef GetInstanceFieldAddress(StackEntry objectEntry, FieldDesc field)
+        {
+            var objectType = objectEntry.Type ?? field.OwningType;
+            LLVMValueRef typedObjectValue;
+            if (objectType.IsValueType && !objectType.IsPointer && objectEntry.Kind != StackValueKind.NativeInt)
+            {
+                typedObjectValue = LLVM.BuildAlloca(_builder, GetLLVMTypeForTypeDesc(objectType), "objptr");
+                LLVM.BuildStore(_builder, objectEntry.LLVMValue, typedObjectValue);
+            }
+            else
+            {
+                typedObjectValue = objectEntry.LLVMValue;
+            }
+
+            var untypedObjectPointer = LLVM.BuildPointerCast(_builder, typedObjectValue, LLVM.PointerType(LLVMTypeRef.Int8Type(), 0), String.Empty);
+            var loadLocation = LLVM.BuildGEP(_builder, untypedObjectPointer,
+            new LLVMValueRef[] { LLVM.ConstInt(LLVM.Int32Type(), (ulong)field.Offset.AsInt, LLVMMisc.False) }, String.Empty);
+            return LLVM.BuildPointerCast(_builder, loadLocation, LLVM.PointerType(GetLLVMTypeForTypeDesc(field.FieldType), 0), String.Empty);
+        }
+
+        private LLVMValueRef GetFieldAddress(FieldDesc field, bool isStatic)
+        {
+            if (field.IsStatic)
+            {
+                //pop unused value
+                if (!isStatic)
+                    _stack.Pop();
+
+                LLVMValueRef untypedFieldAddress = WebAssemblyObjectWriter.EmitGlobal(Module, field, _compilation.NameMangler);
```

Comment:
For static field accesses, we probably also have to deal with running static constructors. You can take a look at https://github.com/dotnet/corert/blob/5705c262df6b6da588e65eb4d7224700e10d9c87/src/ILCompiler.CppCodeGen/src/CppCodeGen/ILToCppImporter.cs#L2677 for inspiration.

Reply:
Can I flag and trigger the cctor on entry to the current routine rather that at the access point? I don't see a way to inject an additional LLVM BasicBlock to deal with the if(!initialized) DoCctor(); without wrecking things.